Author: Reinhard Max <max@suse.de>

XDMCP: For IPv6 add IPv6 link local addresses to the end of the list

For IPv6 add a link local addresses to the end of the list passed to
the XDMCP servers.
Reason: for link local addresses the XDMCP server would need to either
know the interface thru a scope identifier or try all available interfaces.
If they don't this address will fail in which case the XDMCP server
could still try the other addresses passed - however some only try
the first address and then give up.
Even if this seems to be the wrong place to fix this it seems to be
easier than fixing all display servers.

Index: xorg-server-1.20.5/os/access.c
===================================================================
--- xorg-server-1.20.5.orig/os/access.c
+++ xorg-server-1.20.5/os/access.c
@@ -827,7 +827,9 @@ DefineSelf(int fd)
 
             /*
              * ignore 'localhost' entries as they're not useful
-             * on the other end of the wire
+             * on the other end of the wire and because on hosts
+             * with shared home dirs they'll result in conflicting
+             * entries in ~/.Xauthority
              */
             if (ifr->ifa_flags & IFF_LOOPBACK)
                 continue;
@@ -848,6 +850,14 @@ DefineSelf(int fd)
             else if (family == FamilyInternet6 &&
                      IN6_IS_ADDR_LOOPBACK((struct in6_addr *) addr))
                 continue;
+
+      /* Ignore IPv6 link local addresses (fe80::/10), because
+       * they need a scope identifier, which we have no way
+       * of telling to the other end.
+       */
+      if (family == FamilyInternet6 &&
+         IN6_IS_ADDR_LINKLOCAL((struct in6_addr *)addr))
+         continue;
 #endif
             XdmcpRegisterConnection(family, (char *) addr, len);
 #if defined(IPv6) && defined(AF_INET6)
